helix-view/src/document.rs

```diff
@@ -307,6 +307,19 @@ pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
     Ok(())
 }
 
+/// Inserts the final line ending into `rope` if it's missing. [Why?](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline)
+pub fn with_line_ending(rope: &mut Rope) -> LineEnding {
+    // search for line endings
+    let line_ending = auto_detect_line_ending(rope).unwrap_or(DEFAULT_LINE_ENDING);
+
+    // add missing newline at the end of file
+    if rope.len_bytes() == 0 || !char_is_line_ending(rope.char(rope.len_chars() - 1)) {
+        rope.insert(rope.len_chars(), line_ending.as_str());
+    }
+
+    line_ending
+}
+
```

Comment:
I don't think we're using this code anywhere else, and it doesn't strike me that we're likely too.  If it were me, I'd just leave it as it was rather than breaking it out into a function.  But again, won't block the PR on this.

Reply:
Ha, I should read the whole code before commenting.  I see now that it is used elsewhere.  That's some egg on my face!

In that case, though, there are other things to also roll into it, like the indent-style auto-detection, etc.  Basically everything about setting up the document that's not literally loading it from disk.  But that can be cleaned up later, too.

Reply:
Ah, now I remember I was supposed to swap to using an internal `fn open_impl()`, oops.